_drafts/oryx/dependency-injection.md

```diff
@@ -0,0 +1,228 @@
+
+```ts
+export class DefaultCartService implements CartService {
+  // implementation
+}
+```
+
+# Providing services
+
+To use and inject services, they must be provided in the DI container. Oryx offers several types of providers:
+
+- `ClassProvider<T>`: Binds a token to a class constructor.
+- `ValueProvider<T>`: Binds a token to a value.
+- `FactoryProvider<T>`: Binds a token to a factory function.
+- `ExistingProvider<T>`: Binds a token to an existing token.
+
+You can pass a list of providers as an argument to the `Injector` constructor or to the Injector's `provide` method. However, it's important to note that providing services is only possible before the first usage of the DI container.
+
+In Oryx applications, providers are typically grouped into const arrays and passed to the app orchestrator either directly or as part of feature definitions. By doing this, you can easily manage features with corresponding providers and their dependencies.
+
+Here's an example describing how to use the app orchestrator to provide services as part of a feature:
+
+```ts
+export const app = appBuilder()
+  // ...
+  .withFeature({
+    providers: [{ provide: CartService, useClass: CustomizedCartService }],
+  })
+  // ...
+  .create();
+```
+
+In this example, customized cart service is provided by specifying the providers property within a feature object. This way, you can easily manage and extend the services provided by your Oryx application.
+
+# Injecting INJECTOR
+
+In some cases, dependencies may be optional or resolved at runtime, making it difficult to inject them directly into a service. In these situations, you can inject the current DI container using the `INJECTOR` token, like in the following example:
+
+```ts
+constructor(protected injector = inject(INJECTOR)) {}
+```
+
+With the `INJECTOR` token injected, you can use the injector's `inject` method to dynamically resolve dependencies at runtime. For example:
+
+```ts
+someServiceMethod(field) {
+	const template = this.injector.inject(`${FormFieldRenderer}-${field.type}`);
+}
+
+```
+
+In this example, the `inject` method is used to resolve a dependency based on a dynamic token that includes the field type. Since the field type is only known at runtime, it is not possible to inject the dependency directly. Instead, the `INJECTOR` token is injected, allowing the `inject` method to resolve the dependency dynamically.
+
+# Multi Providers
+
+While most dependencies in an app correspond to only one value, such as a class, there are situations where we want to have dependencies with multiple different values, for example HTTP interceptors or normalizers. However, it's not very practical to configure all these dependencies separately, because we want to access them all together at once. Therefore, we can use a special type of dependency that accepts multiple values, not just one, linked to the exact same dependency injection token. These are called multi-providers.
```

Comment:
In the second sentence, replace "we" with what actually needs to access values at once.

Reply:
```suggestion
While most dependencies in an app correspond to only one value, such as a class, there are situations where dependencies with multiple different values are needed, for example HTTP interceptors or normalizers. However, it's not very practical to configure all these dependencies separately, because the system or application wants to access them all together at once. Therefore, a special type of dependency that accepts multiple values, not just one, linked to the exact same dependency injection token can be used. These are called multi-providers.
```